internal/datastore/proxy/caching.go

```diff
@@ -36,72 +43,181 @@ func DatastoreProxyTestCache(t testing.TB) cache.Cache {
 	datastore.Datastore
-	c           cache.Cache
-	readNsGroup singleflight.Group
+	c         cache.Cache
+	readGroup singleflight.Group
 }
 
-func (p *nsCachingProxy) Close() error {
+func (p *definitionCachingProxy) Close() error {
 	p.c.Close()
 	return p.Datastore.Close()
 }
 
-func (p *nsCachingProxy) SnapshotReader(rev datastore.Revision) datastore.Reader {
+func (p *definitionCachingProxy) SnapshotReader(rev datastore.Revision) datastore.Reader {
 	delegateReader := p.Datastore.SnapshotReader(rev)
-	return &nsCachingReader{delegateReader, rev, p}
+	return &definitionCachingReader{delegateReader, rev, p}
 }
 
-func (p *nsCachingProxy) ReadWriteTx(
+func (p *definitionCachingProxy) ReadWriteTx(
 	ctx context.Context,
 	f datastore.TxUserFunc,
 ) (datastore.Revision, error) {
 	return p.Datastore.ReadWriteTx(ctx, func(delegateRWT datastore.ReadWriteTransaction) error {
-		rwt := &nsCachingRWT{delegateRWT, &sync.Map{}}
+		rwt := &definitionCachingRWT{delegateRWT, &sync.Map{}}
 		return f(rwt)
 	})
 }
 
-type nsCachingReader struct {
+type definitionCachingReader struct {
 	datastore.Reader
 	rev datastore.Revision
-	p   *nsCachingProxy
+	p   *definitionCachingProxy
 }
 
-func (r *nsCachingReader) ReadNamespaceByName(
+func (r *definitionCachingReader) ReadNamespaceByName(
 	ctx context.Context,
-	nsName string,
+	name string,
 ) (*core.NamespaceDefinition, datastore.Revision, error) {
-	// Check the nsCache.
-	nsRevisionKey := nsName + "@" + r.rev.String()
+	return readAndCache[*core.NamespaceDefinition](ctx, r, "namespace", name,
+		func(ctx context.Context, name string) (*core.NamespaceDefinition, datastore.Revision, error) {
+			return r.Reader.ReadNamespaceByName(ctx, name)
+		},
+		estimatedNamespaceDefinitionSize)
+}
+
+func (r *definitionCachingReader) LookupNamespacesWithNames(
+	ctx context.Context,
+	nsNames []string,
+) ([]datastore.RevisionedNamespace, error) {
+	return listAndCache[*core.NamespaceDefinition](ctx, r, "namespace", nsNames,
+		func(ctx context.Context, names []string) ([]datastore.RevisionedNamespace, error) {
+			return r.Reader.LookupNamespacesWithNames(ctx, names)
+		},
+		estimatedNamespaceDefinitionSize)
+}
+
+func (r *definitionCachingReader) ReadCaveatByName(
+	ctx context.Context,
+	name string,
+) (*core.CaveatDefinition, datastore.Revision, error) {
+	return readAndCache[*core.CaveatDefinition](ctx, r, "caveat", name,
+		func(ctx context.Context, name string) (*core.CaveatDefinition, datastore.Revision, error) {
+			return r.Reader.ReadCaveatByName(ctx, name)
+		},
+		estimatedCaveatDefinitionSize)
+}
+
+func (r *definitionCachingReader) LookupCaveatsWithNames(
+	ctx context.Context,
+	caveatNames []string,
+) ([]datastore.RevisionedCaveat, error) {
+	return listAndCache[*core.CaveatDefinition](ctx, r, "caveat", caveatNames,
+		func(ctx context.Context, names []string) ([]datastore.RevisionedCaveat, error) {
+			return r.Reader.LookupCaveatsWithNames(ctx, names)
+		},
+		estimatedNamespaceDefinitionSize)
```

Comment:
it's using the wrong function: `estimatedCaveatDefinitionSize`

Reply:
good catch. fixed